{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8ff1b192-9c27-4528-aa70-e54c1d26c803",
   "metadata": {},
   "source": [
    "# Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6a47772b-2f02-4ccd-a1cf-dd893f2d7e0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.utils import to_categorical\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import LabelEncoder, StandardScaler\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import tensorflow as tf\n",
    "import os\n",
    "\n",
    "%matplotlib inline\n",
    "seed = 0\n",
    "np.random.seed(seed)\n",
    "\n",
    "tf.random.set_seed(seed)\n",
    "\n",
    "os.environ['XILINX_VITIS'] = '/tools/Xilinx/Vitis/2023.2'\n",
    "os.environ['PATH'] = os.environ['XILINX_VITIS'] + '/bin:' + os.environ['PATH']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bdee486-f401-47c7-b4c6-39ea6bd04333",
   "metadata": {},
   "source": [
    "# Simulate Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "688ce2a9-a5d7-4fce-b279-3d085f3a700f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -- Parameters --\n",
    "fs         = 378e3        # ADC sampling rate [Hz]\n",
    "bit_res    = 20           # 20-bit resolution\n",
    "I_FS       = 1e-3         # full-scale = ±1 mA\n",
    "DIG_FS     = 52000        # device maps ±1 mA → ±52 000 counts\n",
    "N          = 100_000      # total samples\n",
    "f_sig      = 1e3          # setting sine wave frequency to 1 kHz\n",
    "n_channels = 4            # number of ADC input/output channels\n",
    "t = np.arange(N) / fs     # time scale\n",
    "\n",
    "# same sinusoidal current on each channel\n",
    "I_true = I_FS * np.sin(2 * np.pi* f_sig * t).astype(np.float32)\n",
    "I_true_ch = np.tile(I_true.reshape(-1,1), (1, n_channels)) # duplicates array 4x for 4 channels\n",
    "\n",
    "# each channel's gain/offset/noise\n",
    "gains   = np.array([1.05, 0.98, 1.02, 1.10], dtype=np.float32)   # caste everything as 32-bit FP\n",
    "offsets = np.array([ 100, -200,   50,    0], dtype=np.float32)\n",
    "noise   = np.random.normal(0, 50, size=(N, n_channels)).astype(np.float32)\n",
    "\n",
    "# raw ADC counts = ideal-counts * gain + offset + noise\n",
    "ideal_dig = (I_true_ch / I_FS) * DIG_FS # the ideal digital value (ex. if I_true_ch = 0.5mA, then (0.5 / 1) * 52000 = 26000)\n",
    "X_raw = ideal_dig * gains + offsets + noise # [w0 * x0 + b0 + b1]\n",
    "X_raw = np.clip(X_raw, -DIG_FS, + DIG_FS).astype(np.float32) # saturates values out of range\n",
    "\n",
    "y_counts = ideal_dig.astype(np.float32) # current that we want to predict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "140b41b6-4cbc-42e8-a2c8-9301d7d17e73",
   "metadata": {},
   "source": [
    "# Create Training-Test Split of Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84c36b38-08bb-44ea-b5cb-5185fbfac29f",
   "metadata": {},
   "source": [
    "## 80% Training | 20% Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "d36b6aec-d125-4384-bb8f-1038239ffccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = X_raw\n",
    "y = y_counts         \n",
    "\n",
    "# 20% for testing, 80% for training\n",
    "X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f40228f1-3f11-4eb6-b45f-504adea8d5a7",
   "metadata": {},
   "source": [
    "# Scale data for NN to converge faster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "20b2c845-c729-408a-bfda-34a0106e3c7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: (80000, 4) (80000, 4) (20000, 4) (20000, 4)\n"
     ]
    }
   ],
   "source": [
    "scaler = StandardScaler()\n",
    "X_train_val = scaler.fit_transform(X_train_val)\n",
    "X_test = scaler.transform(X_test)\n",
    "\n",
    "print(\"Shapes:\", X_train_val.shape, y_train_val.shape, X_test.shape, y_test.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1194e44a-cf1d-4f2a-aa0b-5d11b2679cc3",
   "metadata": {},
   "source": [
    "# Construct the Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "332e6020-c333-4aae-b911-f26d12f74982",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.models import Sequential\n",
    "from tensorflow.keras.optimizers import Adam\n",
    "from tensorflow.keras.regularizers import l1\n",
    "from callbacks import all_callbacks\n",
    "from tensorflow.keras.layers import Activation\n",
    "from qkeras.qlayers import QDense, QActivation\n",
    "from qkeras.quantizers import quantized_bits, quantized_relu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f4d70bd0-4f44-42fa-88be-b217dccae319",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"sequential_6\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " calib_lin (QDense)          (None, 4)                 20        \n",
      "                                                                 \n",
      " linear_out (Activation)     (None, 4)                 0         \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 20\n",
      "Trainable params: 20\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "model = Sequential()\n",
    "model.add(\n",
    "    QDense(\n",
    "        n_channels,\n",
    "        input_shape=(n_channels,),\n",
    "        name='calib_lin',\n",
    "        kernel_quantizer=quantized_bits(bit_res, 0, alpha=1),\n",
    "        bias_quantizer  =quantized_bits(bit_res, 0, alpha=1),\n",
    "        kernel_initializer='lecun_uniform'\n",
    "    )\n",
    ")\n",
    "model.add(\n",
    "    Activation(\n",
    "        'linear',\n",
    "        name='linear_out'\n",
    "    )\n",
    ")\n",
    "\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b77f2dc7-a6fc-42eb-a2a5-8bb0f821763d",
   "metadata": {},
   "source": [
    "# Don't meed to train with Sparsity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "dc0ff123-6cfb-4b11-9cc0-919637ce9b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# you dont need to do sparsity for a model this simple"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a52db33-0f00-4fee-8cbe-63a273d08b89",
   "metadata": {},
   "source": [
    "# Train the Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ab4165d8-5ee2-4e6b-9292-b7ec732c21f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:`epsilon` argument is deprecated and will be removed, use `min_delta` instead.\n",
      "WARNING:tensorflow:`period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.\n",
      "Epoch 1/50\n",
      "  1/250 [..............................] - ETA: 49s - loss: 1356996480.0000 - mae: 33213.9141WARNING:tensorflow:Callback method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0006s vs `on_train_batch_end` time: 0.0013s). Check your callbacks.\n",
      "237/250 [===========================>..] - ETA: 0s - loss: 1353591424.0000 - mae: 33123.7812\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 1: val_loss improved from inf to 1347736960.00000, saving model to model_1/KERAS_check_best_model.h5\n",
      "\n",
      "Epoch 1: val_loss improved from inf to 1347736960.00000, saving model to model_1/KERAS_check_best_model_weights.h5\n",
      "\n",
      "Epoch 1: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 1: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 1s 2ms/step - loss: 1352429184.0000 - mae: 33109.5508 - val_loss: 1347736960.0000 - val_mae: 33039.3242 - lr: 1.0000e-04\n",
      "Epoch 2/50\n",
      "182/250 [====================>.........] - ETA: 0s - loss: 1354207360.0000 - mae: 33149.0742\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 2: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 2: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 2: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 2: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429312.0000 - mae: 33109.5430 - val_loss: 1347736960.0000 - val_mae: 33039.3242 - lr: 1.0000e-04\n",
      "Epoch 3/50\n",
      "218/250 [=========================>....] - ETA: 0s - loss: 1350581248.0000 - mae: 33077.1992\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 3: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 3: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 3: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 3: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 988us/step - loss: 1352429312.0000 - mae: 33109.5469 - val_loss: 1347736960.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 4/50\n",
      "213/250 [========================>.....] - ETA: 0s - loss: 1352670848.0000 - mae: 33115.9922\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 4: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 4: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 4: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 4: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429440.0000 - mae: 33109.5352 - val_loss: 1347736960.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 5/50\n",
      "215/250 [========================>.....] - ETA: 0s - loss: 1351472896.0000 - mae: 33094.1719\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 5: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 5: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 5: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 5: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429184.0000 - mae: 33109.5586 - val_loss: 1347736960.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 6/50\n",
      "221/250 [=========================>....] - ETA: 0s - loss: 1352979712.0000 - mae: 33113.5547\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 6: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 6: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 6: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 6: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 855us/step - loss: 1352429056.0000 - mae: 33109.5586 - val_loss: 1347736960.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 7/50\n",
      "178/250 [====================>.........] - ETA: 0s - loss: 1353014784.0000 - mae: 33118.2422\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 7: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 7: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 7: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 7: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429056.0000 - mae: 33109.5664 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 8/50\n",
      "231/250 [==========================>...] - ETA: 0s - loss: 1354175744.0000 - mae: 33139.9844\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 8: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 8: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 8: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 8: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352428800.0000 - mae: 33109.5508 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 9/50\n",
      "214/250 [========================>.....] - ETA: 0s - loss: 1352816128.0000 - mae: 33110.2695\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 9: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 9: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 9: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 9: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352428544.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 10/50\n",
      "185/250 [=====================>........] - ETA: 0s - loss: 1355080832.0000 - mae: 33145.3086\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 10: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 10: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 10: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 10: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 10: saving model to model_1/KERAS_check_model_epoch10.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 992us/step - loss: 1352428928.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 11/50\n",
      "202/250 [=======================>......] - ETA: 0s - loss: 1353362304.0000 - mae: 33127.9141\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 11: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 11: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 11: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 11: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 11: ReduceLROnPlateau reducing learning rate to 4.999999873689376e-05.\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 695us/step - loss: 1352428800.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 1.0000e-04\n",
      "Epoch 12/50\n",
      "209/250 [========================>.....] - ETA: 0s - loss: 1350721792.0000 - mae: 33078.4297\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 12: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 12: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 12: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 12: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 986us/step - loss: 1352429184.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 13/50\n",
      "222/250 [=========================>....] - ETA: 0s - loss: 1352504320.0000 - mae: 33107.2656\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 13: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 13: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 13: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 13: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 883us/step - loss: 1352429696.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 14/50\n",
      "239/250 [===========================>..] - ETA: 0s - loss: 1352146304.0000 - mae: 33100.7852\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 14: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 14: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 14: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 14: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 2ms/step - loss: 1352429312.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 15/50\n",
      "247/250 [============================>.] - ETA: 0s - loss: 1352438656.0000 - mae: 33111.0352\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 15: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 15: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 15: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 15: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352428544.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 16/50\n",
      "241/250 [===========================>..] - ETA: 0s - loss: 1351266176.0000 - mae: 33090.0078\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 16: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 16: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 16: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 16: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 823us/step - loss: 1352429824.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 17/50\n",
      "237/250 [===========================>..] - ETA: 0s - loss: 1352535936.0000 - mae: 33110.5117\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 17: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 17: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 17: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 17: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352428416.0000 - mae: 33109.5586 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 18/50\n",
      "161/250 [==================>...........] - ETA: 0s - loss: 1350362368.0000 - mae: 33065.5938\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 18: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 18: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 18: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 18: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 805us/step - loss: 1352429312.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 19/50\n",
      "200/250 [=======================>......] - ETA: 0s - loss: 1355022592.0000 - mae: 33153.9219\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 19: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 19: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 19: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 19: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 691us/step - loss: 1352429824.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 5.0000e-05\n",
      "Epoch 20/50\n",
      "179/250 [====================>.........] - ETA: 0s - loss: 1353117312.0000 - mae: 33113.1719\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 20: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 20: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 20: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 20: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 20: saving model to model_1/KERAS_check_model_epoch20.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 848us/step - loss: 1352429440.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 21/50\n",
      "204/250 [=======================>......] - ETA: 0s - loss: 1352590720.0000 - mae: 33111.8750\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 21: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 21: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 21: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 21: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 946us/step - loss: 1352428672.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3281 - lr: 5.0000e-05\n",
      "Epoch 22/50\n",
      "191/250 [=====================>........] - ETA: 0s - loss: 1350857216.0000 - mae: 33077.7695\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 22: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 22: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 22: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 22: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 22: ReduceLROnPlateau reducing learning rate to 2.499999936844688e-05.\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 738us/step - loss: 1352429568.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 5.0000e-05\n",
      "Epoch 23/50\n",
      "249/250 [============================>.] - ETA: 0s - loss: 1352391296.0000 - mae: 33109.0273\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 23: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 23: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 23: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 23: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 863us/step - loss: 1352428928.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 24/50\n",
      "225/250 [==========================>...] - ETA: 0s - loss: 1354526592.0000 - mae: 33135.1016\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 24: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 24: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 24: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 24: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352428800.0000 - mae: 33109.5391 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 25/50\n",
      "155/250 [=================>............] - ETA: 0s - loss: 1349419136.0000 - mae: 33077.8477\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 25: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 25: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 25: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 25: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 855us/step - loss: 1352429056.0000 - mae: 33109.5391 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 26/50\n",
      "193/250 [======================>.......] - ETA: 0s - loss: 1356765184.0000 - mae: 33192.6641\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 26: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 26: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 26: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 26: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 725us/step - loss: 1352428672.0000 - mae: 33109.5508 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 27/50\n",
      "192/250 [======================>.......] - ETA: 0s - loss: 1354447360.0000 - mae: 33138.1992\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 27: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 27: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 27: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 27: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 932us/step - loss: 1352429056.0000 - mae: 33109.5508 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 28/50\n",
      "207/250 [=======================>......] - ETA: 0s - loss: 1351328512.0000 - mae: 33079.7266\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 28: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 28: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 28: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 28: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352428928.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 29/50\n",
      "189/250 [=====================>........] - ETA: 0s - loss: 1353077760.0000 - mae: 33118.1641\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 29: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 29: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 29: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 29: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 701us/step - loss: 1352429312.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 30/50\n",
      "231/250 [==========================>...] - ETA: 0s - loss: 1352303616.0000 - mae: 33098.7969\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 30: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 30: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 30: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 30: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 30: saving model to model_1/KERAS_check_model_epoch30.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429952.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 31/50\n",
      "215/250 [========================>.....] - ETA: 0s - loss: 1349804416.0000 - mae: 33072.3477\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 31: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 31: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 31: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 31: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429184.0000 - mae: 33109.5508 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 32/50\n",
      "233/250 [==========================>...] - ETA: 0s - loss: 1352513536.0000 - mae: 33106.0469\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 32: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 32: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 32: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 32: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429056.0000 - mae: 33109.5352 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 33/50\n",
      "186/250 [=====================>........] - ETA: 0s - loss: 1351243136.0000 - mae: 33096.7500\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 33: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 33: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 33: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 33: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 33: ReduceLROnPlateau reducing learning rate to 1.249999968422344e-05.\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 757us/step - loss: 1352428928.0000 - mae: 33109.5508 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 2.5000e-05\n",
      "Epoch 34/50\n",
      "199/250 [======================>.......] - ETA: 0s - loss: 1352443648.0000 - mae: 33109.6055\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 34: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 34: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 34: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 34: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429568.0000 - mae: 33109.5508 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 35/50\n",
      "221/250 [=========================>....] - ETA: 0s - loss: 1351132032.0000 - mae: 33083.6680\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 35: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 35: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 35: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 35: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 921us/step - loss: 1352429312.0000 - mae: 33109.5586 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 36/50\n",
      "235/250 [===========================>..] - ETA: 0s - loss: 1352209664.0000 - mae: 33110.0547\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 36: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 36: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 36: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 36: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429568.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 37/50\n",
      "223/250 [=========================>....] - ETA: 0s - loss: 1352170624.0000 - mae: 33108.7148\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 37: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 37: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 37: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 37: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429696.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 38/50\n",
      "206/250 [=======================>......] - ETA: 0s - loss: 1354098176.0000 - mae: 33128.3125\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 38: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 38: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 38: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 38: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429440.0000 - mae: 33109.5508 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 39/50\n",
      "184/250 [=====================>........] - ETA: 0s - loss: 1349173632.0000 - mae: 33066.3242\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 39: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 39: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 39: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 39: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429824.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 40/50\n",
      "239/250 [===========================>..] - ETA: 0s - loss: 1352878080.0000 - mae: 33122.5039\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 40: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 40: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 40: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 40: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 40: saving model to model_1/KERAS_check_model_epoch40.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352428928.0000 - mae: 33109.5391 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 41/50\n",
      "223/250 [=========================>....] - ETA: 0s - loss: 1352695808.0000 - mae: 33107.8281\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 41: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 41: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 41: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 41: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 881us/step - loss: 1352429440.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 42/50\n",
      "223/250 [=========================>....] - ETA: 0s - loss: 1350362240.0000 - mae: 33076.7578\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 42: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 42: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 42: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 42: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429568.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 43/50\n",
      "221/250 [=========================>....] - ETA: 0s - loss: 1353978112.0000 - mae: 33134.1875\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 43: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 43: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 43: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 43: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429440.0000 - mae: 33109.5391 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 44/50\n",
      "230/250 [==========================>...] - ETA: 0s - loss: 1352515328.0000 - mae: 33110.7969\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 44: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 44: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 44: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 44: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 44: ReduceLROnPlateau reducing learning rate to 6.24999984211172e-06.\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429056.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 1.2500e-05\n",
      "Epoch 45/50\n",
      "230/250 [==========================>...] - ETA: 0s - loss: 1353312128.0000 - mae: 33121.8984\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 45: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 45: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 45: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 45: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 916us/step - loss: 1352429696.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 6.2500e-06\n",
      "Epoch 46/50\n",
      "234/250 [===========================>..] - ETA: 0s - loss: 1351916160.0000 - mae: 33100.3164\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 46: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 46: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 46: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 46: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 1ms/step - loss: 1352429184.0000 - mae: 33109.5547 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 6.2500e-06\n",
      "Epoch 47/50\n",
      "192/250 [======================>.......] - ETA: 0s - loss: 1350124032.0000 - mae: 33085.2930\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 47: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 47: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 47: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 47: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 738us/step - loss: 1352429440.0000 - mae: 33109.5391 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 6.2500e-06\n",
      "Epoch 48/50\n",
      "185/250 [=====================>........] - ETA: 0s - loss: 1350897664.0000 - mae: 33094.3672\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 48: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 48: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 48: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 48: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 763us/step - loss: 1352429184.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 6.2500e-06\n",
      "Epoch 49/50\n",
      "185/250 [=====================>........] - ETA: 0s - loss: 1352560384.0000 - mae: 33114.4336\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 49: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 49: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 49: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 49: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 748us/step - loss: 1352429056.0000 - mae: 33109.5469 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 6.2500e-06\n",
      "Epoch 50/50\n",
      "192/250 [======================>.......] - ETA: 0s - loss: 1349965696.0000 - mae: 33071.1133\n",
      "***callbacks***\n",
      "saving losses to model_1/losses.log\n",
      "\n",
      "Epoch 50: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 50: val_loss did not improve from 1347736960.00000\n",
      "\n",
      "Epoch 50: saving model to model_1/KERAS_check_model_last.h5\n",
      "\n",
      "Epoch 50: saving model to model_1/KERAS_check_model_last_weights.h5\n",
      "\n",
      "Epoch 50: saving model to model_1/KERAS_check_model_epoch50.h5\n",
      "\n",
      "***callbacks end***\n",
      "\n",
      "250/250 [==============================] - 0s 734us/step - loss: 1352429184.0000 - mae: 33109.5430 - val_loss: 1347737088.0000 - val_mae: 33039.3242 - lr: 6.2500e-06\n"
     ]
    }
   ],
   "source": [
    "train = True\n",
    "if train:\n",
    "    model.compile(\n",
    "        optimizer=Adam(learning_rate=1e-4),\n",
    "        loss='mse',\n",
    "        metrics=['mae']\n",
    "    )\n",
    "\n",
    "      # 3) Build your standard callbacks\n",
    "    callbacks = all_callbacks(\n",
    "        stop_patience=1000,\n",
    "        lr_factor=0.5,\n",
    "        lr_patience=10,\n",
    "        lr_epsilon=1e-6,\n",
    "        lr_cooldown=2,\n",
    "        lr_minimum=1e-7,\n",
    "        outputDir='model_1',\n",
    "    )\n",
    "\n",
    "    # 4) Fit with pruning enabled\n",
    "    history = model.fit(\n",
    "        X_train_val, \n",
    "        y_counts,\n",
    "        batch_size=256,\n",
    "        epochs=30,\n",
    "        validation_split=0.2,\n",
    "        shuffle=True,\n",
    "        callbacks=callbacks.callbacks\n",
    "    )\n",
    "\n",
    "    model.save('model_1/calibrated_adc.h5')\n",
    "\n",
    "else:\n",
    "    from tensorflow.keras.models import load_model\n",
    "    model = load_model('model_1/calibrated_adc.h5')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bce47b64-e6e1-4b71-aa42-65fd9b4ef6f4",
   "metadata": {},
   "source": [
    "# Evaluation Metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "80970c6a-0966-49d8-9586-c5cf566a160b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Test MAE = 310889.3 µA\n"
     ]
    }
   ],
   "source": [
    "loss, mae = model.evaluate(X_test, y_test, verbose=0)\n",
    "print(f\"\\nTest MAE = {mae*1e6:.1f} µA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ab230df-de1c-450f-9c61-9fcb3a2275a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = model.predict(X_test)\n",
    "\n",
    "plt.figure(figsize=(5,5))\n",
    "plt.scatter(y_test.flatten(), y_pred.flatten(), s=5, alpha=0.3)\n",
    "lim = CODE_FS / CODE_FS * I_FS\n",
    "plt.plot([-I_FS,I_FS], [-I_FS,I_FS], 'r--')\n",
    "plt.xlabel(\"True current (A)\")\n",
    "plt.ylabel(\"Predicted current (A)\")\n",
    "plt.title(\"Calibration: Pred vs True\")\n",
    "plt.axis('equal')\n",
    "plt.grid(True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81e3c8c5-4d9d-426f-9a2c-228e52515470",
   "metadata": {},
   "source": [
    "# Load Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b107d67b-b5b9-4d8c-8fcb-7f6d8f502f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.models import load_model\n",
    "from qkeras.utils import _add_supported_quantized_objects\n",
    "\n",
    "co = {}\n",
    "_add_supported_quantized_objects(co)\n",
    "import os\n",
    "\n",
    "os.environ['XILINX_VIVADO'] = '/tools/Xilinx/Vivado/2019.2'\n",
    "os.environ['PATH'] = os.environ['XILINX_VIVADO'] + '/bin:' + os.environ['PATH']\n",
    "\n",
    "model = load_model('model_1/calibrated_adc.h5', custom_objects=co)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5da8a1d-d41f-4a47-a393-a165608ea0c8",
   "metadata": {},
   "source": [
    "# Convert to hls4ml model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28a05c38-34d0-41d0-b249-05135beed285",
   "metadata": {},
   "outputs": [],
   "source": [
    "import hls4ml\n",
    "\n",
    "config = hls4ml.utils.config_from_keras_model(model, granularity='name')\n",
    "config['LayerName']['calib_lin']['ReuseFactor'] = 64\n",
    "\n",
    "print(\"-----------------------------------\")\n",
    "plotting.print_dict(config)\n",
    "print(\"-----------------------------------\")\n",
    "hls_model = hls4ml.converters.convert_from_keras_model(\n",
    "    model, hls_config=config, output_dir='model_1/hls4ml_prj_calibrated_adc', backend='VivadoAccelerator', board='pynq-z2'\n",
    ")\n",
    "hls_model.compile()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89ccbb19-ca15-484b-95e5-2360a3038aad",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotting.print_dict(hls4ml.backends.get_backend('VivadoAccelerator').create_initial_config())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22c51c98-5584-4569-8f4f-22d567d86e07",
   "metadata": {},
   "source": [
    "# CPU Simulation of hls4ml model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e79a94ad-9a46-4854-8f72-1a4c2761371b",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_test_c = np.ascontiguousarray(X_test)\n",
    "\n",
    "# Run the HLS model in CPU (C-simulation) mode\n",
    "y_hls = hls_model.predict(X_test_c)\n",
    "\n",
    "# Save the HLS outputs for later comparison \n",
    "output_path = 'model_1/y_hls.npy'\n",
    "np.save(output_path, y_hls)\n",
    "print(f\"HLS C-sim outputs saved to {output_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0fba87d-fa1d-4cf1-814e-e6ec6473ffea",
   "metadata": {},
   "source": [
    "# Synthesize & Make Bitfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f97cdbc-73a1-418f-893e-be4dc47737c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "hls_model.build(csim=False, synth=True, export=True, bitfile=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "585f4087-4202-46f8-aaaa-27064ca71ac7",
   "metadata": {},
   "source": [
    "# Final Resource Usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58a72ca1-818c-4704-8ab1-1934b5912f2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed -n '30,45p' model_1/hls4ml_prj_pynq/myproject_vivado_accelerator/project_1.runs/impl_1/design_1_wrapper_utilization_placed.rpt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
